File: solune/frontend/src/types/tools.ts
/**
 * MCP Tools domain types (027-mcp-tools-page).
 */

export type McpToolSyncStatus = 'synced' | 'pending' | 'error';

export interface McpToolConfig {
  id: string;
  name: string;
  description: string;
  endpoint_url: string;
  config_content: string;
  sync_status: McpToolSyncStatus;
  sync_error: string;
  synced_at: string | null;
  github_repo_target: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface McpToolConfigCreate {
  name: string;
  description: string;
  config_content: string;
  github_repo_target: string;
}

export interface McpToolConfigUpdate {
  name?: string;
  description?: string;
  config_content?: string;
  github_repo_target?: string;
}

export interface McpToolConfigListResponse {
  tools: McpToolConfig[];
  count: number;
}

export interface McpToolSyncResult {
  id: string;
  sync_status: McpToolSyncStatus;
  sync_error: string;
  synced_at: string | null;
  synced_paths: string[];
}

export interface RepoMcpServerConfig {
  name: string;
  config: Record<string, unknown>;
  source_paths: string[];
}

export interface RepoMcpServerUpdate {
  name: string;
  config_content: string;
}

export interface RepoMcpConfigResponse {
  paths_checked: string[];
  available_paths: string[];
  primary_path: string | null;
  servers: RepoMcpServerConfig[];
}

export interface McpPreset {
  id: string;
  name: string;
  description: string;
  category: string;
  icon: string;
  config_content: string;
}

export interface McpPresetListResponse {
  presets: McpPreset[];
  count: number;
}

export interface ToolChip {
  id: string;
  name: string;
  description: string;
}

export interface ToolDeleteResult {
  success: boolean;
  deleted_id: string | null;
  affected_agents: ToolChip[];
}

File: solune/backend/src/models/tools.py
"""Pydantic models for MCP tool configuration management."""

from pydantic import BaseModel, Field


class McpToolConfig(BaseModel):
    """Full MCP tool configuration entity."""

    id: str
    github_user_id: str
    project_id: str
    name: str
    description: str = ""
    endpoint_url: str
    config_content: str = "{}"
    sync_status: str = "pending"
    sync_error: str = ""
    synced_at: str | None = None
    github_repo_target: str = ""
    is_active: bool = True
    created_at: str
    updated_at: str


class McpToolConfigCreate(BaseModel):
    """Request body for creating/uploading a new MCP tool configuration."""

    name: str = Field(min_length=1, max_length=100)
    description: str = Field(default="", max_length=500)
    config_content: str = Field(min_length=2, max_length=262144)
    github_repo_target: str = Field(default="", max_length=200)


class McpToolConfigUpdate(BaseModel):
    """Request body for updating an existing MCP tool configuration."""

    name: str | None = None
    description: str | None = None
    config_content: str | None = None
    github_repo_target: str | None = None


class McpToolConfigResponse(BaseModel):
    """Single MCP tool configuration in API responses."""

    id: str
    name: str
    description: str
    endpoint_url: str
    config_content: str
    sync_status: str
    sync_error: str
    synced_at: str | None
    github_repo_target: str
    is_active: bool
    created_at: str
    updated_at: str


class McpToolConfigListResponse(BaseModel):
    """Response for the list endpoint."""

    tools: list[McpToolConfigResponse]
    count: int


class McpToolConfigSyncResult(BaseModel):
    """Response for sync operations."""

    id: str
    sync_status: str
    sync_error: str
    synced_at: str | None
    synced_paths: list[str] = Field(default_factory=list)


class RepoMcpServerUpdate(BaseModel):
    """Request body for updating a repository MCP server entry directly."""

    name: str = Field(min_length=1, max_length=100)
    config_content: str = Field(min_length=2, max_length=262144)


class RepoMcpServerConfig(BaseModel):
    """A normalized MCP server entry discovered in the repository."""

    name: str
    config: dict[str, object]
    source_paths: list[str] = Field(default_factory=list)


class RepoMcpConfigResponse(BaseModel):
    """Repository MCP configuration discovered from known GitHub paths."""

    paths_checked: list[str] = Field(default_factory=list)
    available_paths: list[str] = Field(default_factory=list)
    primary_path: str | None = None
    servers: list[RepoMcpServerConfig] = Field(default_factory=list[RepoMcpServerConfig])


class McpPresetResponse(BaseModel):
    """Static MCP preset returned to the frontend."""

    id: str
    name: str
    description: str
    category: str
    icon: str
    config_content: str


class McpPresetListResponse(BaseModel):
    """Response for listing available MCP presets."""

    presets: list[McpPresetResponse]
    count: int


class AgentToolAssociation(BaseModel):
    """Represents the many-to-many relationship between agents and MCP tools."""

    agent_id: str
    tool_id: str
    assigned_at: str


class AgentToolInfo(BaseModel):
    """Lightweight tool info returned in agent-tool association endpoints."""

    id: str
    name: str
    description: str


class AgentToolsResponse(BaseModel):
    """Response for agent tools endpoints."""

    tools: list[AgentToolInfo]


class AgentToolsUpdate(BaseModel):
    """Request body for updating agent tool associations."""

    tool_ids: list[str]


class ToolDeleteResult(BaseModel):
    """Response for delete operations."""

    success: bool
    deleted_id: str | None = None
    affected_agents: list[AgentToolInfo] = Field(default_factory=list[AgentToolInfo])


# ── MCP Catalog Models ──


class CatalogInstallConfig(BaseModel):
    """Normalized representation of upstream Glama install instructions."""

    transport: str
    url: str | None = None
    command: str | None = None
    args: list[str] = Field(default_factory=list)
    env: dict[str, object] = Field(default_factory=dict)
    headers: dict[str, object] = Field(default_factory=dict)
    tools: list[str] = Field(default_factory=list)


class CatalogMcpServer(BaseModel):
    """A transient API model for one external MCP server from the Glama catalog."""

    id: str
    name: str
    description: str
    repo_url: str | None = None
    category: str | None = None
    server_type: str
    install_config: CatalogInstallConfig
    quality_score: str | None = None
    already_installed: bool = False


class CatalogMcpServerListResponse(BaseModel):
    """Response for the catalog browse endpoint."""

    servers: list[CatalogMcpServer]
    count: int
    query: str | None = None
    category: str | None = None


class ImportCatalogMcpRequest(BaseModel):
    """Request body for importing a catalog server into a project."""

    catalog_server_id: str = Field(min_length=1)
